### DISCUSSION TITLE
Stuttering during gameplay on an AMD windows machine

### DISCUSSION THREAD

**Discussion by turrrret** [ppy/osu#26998] [Q&A]
### Type

Performance

### Bug description

Hi!
After installing the latest version of osu lazer on my machine (5800x3d, 6800xt), i experience stutters during gameplay. They appear sporadically, but frequently (multiple times per map). I have imported my old osu data through the wizard, chose my skin and changed basic settings like background dim. The windows install is rather recent (2 months old, Windows 10). I am on the directx renderer still.

The stutters come in different forms:

- Microstutters visual
- Audio stops for a split second
- Larger freezes (~0.5s, render time shoots up 20ms+ when looking at the fps counter)

Fixes I've tried:

- Update Graphics Drivers
- Disabled Tablet
- Different FPS limits
- Restarted PC
- Different Skins, Maps
- Different AMD graphic options (anti-lag, etc.. currently all off)

I am running out of ideas though. It's also not loading a map for the first time. It happens on restarts too. Attatched are the logs and a video demonstrating both audio and visual stutter. The extra added logs were generated right after the play that is attached was done.

### Screenshots or videos

https://github.com/ppy/osu/assets/125567133/76e73014-6f74-40f5-9bb1-543e1476912e

### Version

2024.131.0

### Logs

[compressed-logs.zip](https://github.com/ppy/osu/files/14152075/compressed-logs.zip)
[1706899924.performance.log](https://github.com/ppy/osu/files/14152076/1706899924.performance.log)
[1706899924.runtime.log](https://github.com/ppy/osu/files/14152077/1706899924.runtime.log)
[1706899924.updater.log](https://github.com/ppy/osu/files/14152078/1706899924.updater.log)
[1706899924.input.log](https://github.com/ppy/osu/files/14152079/1706899924.input.log)
[1706899924.network.log](https://github.com/ppy/osu/files/14152080/1706899924.network.log)

**Comment by bdach** [comment]
please attach video of this with frame graph showing (ctrl-f11 twice)

**Comment by turrrret** [comment]
https://github.com/ppy/osu/assets/125567133/563317c5-64d8-4b16-9d59-0eaec29bf732

Here is another example. Right at the start and during the spinner.

**Reply by bdach** [reply]
No specific advice I can offer sorry. We simply need to optimise further.

**Comment by turrrret** [comment]
If it helps, my second machine (i9 9900k + 3060Ti) works flawlessly. Seems to be some AMD thing.

**Reply by BlockHead007** [reply]
I have laptop with 5600h and 3060 gpu and don't have any stutters, but sometimes i have freezes like you've described, but i'm getting them not only in osu so i think this is not their fault but AMD's.

**Reply by turrrret** [reply]
Well, for me other games are fine. So is stable

**Comment by mixalisnew** [comment]
i also use an amd cpu and have experienced some stutter issues i will see if i can get some with logs and stuff and report anything more

**Comment by Lamunir** [comment]
I also have the same stuttering issues, AMD 5950x and Nvidia 3080.

**Reply by Lamunir** [reply]
Changing my Nvidia Low Latency settings from “Ultra” to “On” seemed to fix this for me.
It also fixed it on my Intel machine.
14700k + 4080 Super.

**Comment by peppy** [comment]
Does turning off "high precision input mouse" with this?

![2024-02-19 11 12 26@2x](https://github.com/ppy/osu/assets/191335/d942d4b3-1c36-4cb6-b39d-03ab0f2f5e9e)

**Reply by turrrret** [reply]
I have tried turning this setting off, with no effect sadly. I was using tablet, but i can try again with mouse.

I did a fresh windows install a day ago, so i will have to reinstall and see if the issue is still present even

**Reply by turrrret** [reply]
After properly trying again on the new windows install, the same issue persists. Turning high precision mouse off did not change the behavior.

**Comment by trustytrojan** [comment]
I also have a full AMD machine (albeit way older than yours) and I also get visual micro-stuttering when I occasionally play Windows. After playing on Linux for several months now, I haven't seen a single stutter (or haven't seen enough to complain). I think your issue might just be some random software running in the background that might be taking up CPU time. Considering your install is recent, the culprit might be some Microsoft services that you can't do much about. I suggest trying out [AtlasOS](https://atlasos.net/) to see if that improves anything.

**Reply by trustytrojan** [reply]
I'm a mania player myself, and I have no complaints about my experience on Linux. For context my GPU is an R9 380 and CPU is an FX-6300. Decade-old hardware. Maybe the reason I don't see stutters is that the Linux kernel's driver for older AMD cards is more mature. I'm not sure how it's going for newer cards though.

**Reply by tsunyoku** [reply]
for the love of god please do not suggest that people install a modified windows OS that at best removes security features in the OS and at worst does not do what it says it does

**Comment by smoogipoo** [comment]
This may be a re-ocurrence of https://github.com/ppy/osu/issues/11800

Even though it's closed, we never really fixed the issue. And the underlying issue for Linux is still open: https://github.com/dotnet/runtime/issues/76290

I believe .NET 7 or 8 started using GC regions by default. I believe it's turned off on macOS for the time being, but I'm not 100% sure on that. This could be likely culprit, or rather an exacerbation of the above issue.

Can those that are able to reliably reproduce between startups try the following two methods **separately and together** and seeing how the behaviour changes:

1. Set environment variable `DOTNET_GCGen0MaxBudget=600000` (or lower values like `100000`).
2. Set environment variable `DOTNET_GCName=clrgc.dll` (windows) / `DOTNET_GCName=libclrgc.so` (linux) / `DOTNET_GCName=libclrgc.dylib` (macOS).

**Reply by respektive** [reply]
I'm on linux and have recently been getting these stutters. I tried setting the `DOTNET_GCGen0MaxBudget` env var to `100000` and it makes it a lot better. I didn't notice a single stutter in about 30min of playing so far (I was getting at least one per map before). Setting `DOTNET_GCName=libclrgc.so` doesn't seem to have any effect for me.

**Reply by stanriders** [reply]
Setting `DOTNET_GCGen0MaxBudget` to `600000` seems to be removing the stutters almost completely (there are some occasional frame drops but nothing major)

Windows 10, Ryzen 5800x (non-3d)

**Comment by smoogipoo** [comment]
For those that are a bit more experienced, it would be interesting to have a GC trace...

With .NET 8 SDK installed, install the `dotnet-trace` tool:
```
dotnet trace collect -p <process-id> --profile gc-collect
```

Play for a bit to reproduce a few spikes, then Ctrl+C and upload the `.nettrace` file.

**Reply by miandru** [reply]
[osu!.exe_20240303_235535.nettrace.zip](https://github.com/ppy/osu/files/14474960/osu.exe_20240303_235535.nettrace.zip)
Here's a trace for you, it's a very short session, has a few stutters/spikes. Been experiencing this for the past month while trying lazer.

**Comment by respektive** [comment]
Here's a trace that should contain some spikes. 
[osu!_20240226_123145.nettrace.zip](https://github.com/ppy/osu/files/14404444/osu._20240226_123145.nettrace.zip)

meant to reply to https://github.com/ppy/osu/discussions/26998#discussioncomment-8589680, oops

**Comment by smoogipoo** [comment]
The latest release should improve things here.

**Reply by peppy** [reply]
Your vsync issue sounds like something else is interfering with the game, maybe an overlay or maybe a GPU setting. Any kind of "back one frame" is generally something external to osu!.

**Reply by turrrret** [reply]
Hi, as the author of the thread i quickly wanted to give feedback. Things have improved greatly, no more stuttering during gameplay for  me. Thanks!
